lib/galaxy/webapps/galaxy/services/datasets.py

```diff
@@ -606,7 +610,15 @@ def display(
         headers = {}
         rval: Any = ""
         try:
-            dataset_instance = self.dataset_manager_by_type[hda_ldda].get_accessible(dataset_id, trans.user)
+            dataset_instance: Union[model.HistoryDatasetAssociation, model.LibraryDatasetDatasetAssociation]
+            if hda_ldda == "hda":
+                hda_manager = self.dataset_manager_by_type[hda_ldda]
+                dataset_instance = hda = hda_manager.get_accessible(dataset_id, trans.user)
+                hda_manager.ensure_dataset_on_disk(trans, hda)
+            else:
+                ldda_manager = self.dataset_manager_by_type[hda_ldda]
+                dataset_instance = ldda = ldda_manager.get_accessible(dataset_id, trans.user)
+                ldda_manager.ensure_dataset_on_disk(trans, ldda)
```

Comment:
Why not something like this?
```suggestion
            dataset_instance: Union[model.HistoryDatasetAssociation, model.LibraryDatasetDatasetAssociation]
            manager = self.dataset_manager_by_type[hda_ldda]
            dataset_instance = manager.get_accessible(dataset_id, trans.user)
            manager.ensure_dataset_on_disk(trans, dataset_instance)
```

Reply:
I think it's a covariance issue. I agree this is much cleaner, but mypy will complain that either the HDA or the LDDA manager can't work with the oppposite half of the HDA/LDDA union.

Reply:
Ahh I see now... what a pitty 😕